[Draft] Move specification from ROS wiki to the urdfdom repo (#183)
ros/urdfdom: specification/joint.md

# `<joint>` element

The joint element describes the kinematics and dynamics of the joint and also specifies the [safety limits](http://wiki.ros.org/pr2_controller_manager/safety_limits) of the joint.

![example-joint](./joint.png)

## Attributes

| attribute | type        | use      | default value | description                                                                    |
| --------- | ----------- | -------- | ------------- | ------------------------------------------------------------------------------ |
| `name`    | `string`      | required | NA            | Specifies a unique name of the join.                                           |
| `type`    | `JointType` | required | NA            | Specifies the type of joint, see "Type: `JointType`" section for more details. |

### Type: `JointType`

where type can be one of the following:
- `revolute`: a hinge joint that rotates along the axis and has a limited range specified by the upper and lower limits.
- `continuous`: a continuous hinge joint that rotates around the axis and has no upper and lower limits.

> - `prismatic`: a sliding joint that slides along the axis, and has a limited range specified by the upper and lower limits.

Reviewer: ```suggestion
- `prismatic`: a sliding joint that slides along the axis, and has a limited range specified by the upper and lower limits. The joint can only slide along the axis and cannot rotate along it (imagine a rectangular piston, not a cylindrical one).
```

Author: @scpeters @sloretz let me know if you like this change in, thanks!

> - `fixed`: this is not really a joint because it cannot move. All degrees of freedom are locked. This type of joint does not require the `<axis>`, `<calibration>`, `<dynamics>`, `<limits>` or `<safety_controller>`.

Reviewer: ```suggestion
- `fixed`: this is not really a joint because it cannot move. All degrees of freedom are locked. This type of joint does not require the `<axis>`, `<calibration>`, `<dynamics>`, `<limits>` or `<safety_controller>`. It is used to rigidly connect logical parts of the robot.
```

Author: @scpeters @sloretz let me know if you like this change in, thanks!

- `floating`: this joint allows motion for all 6 degrees of freedom.

> - `planar`: this joint allows motion in a plane perpendicular to the axis.

Reviewer: ```suggestion
- `planar`: this joint allows motion in a plane perpendicular to the axis and intersecting the joint frame origin.
```

Author: @scpeters @sloretz let me know if you like this change in, thanks!

| --------- | -------- | -------- | ------------- | --------------------------------------------------------------------------------- |
| `link`    | `string` | required | NA            | The name of the link that is the parent of this link in the robot tree structure. |

### `<child>`

| attribute | type   | use      | default value | description                                  |
| --------- | ------ | -------- | ------------- | -------------------------------------------- |
| `link`    | `string` | required | NA            | The name of the link that is the child link. |

### `<axis>`

The joint axis specified in the joint frame. This is the axis of rotation for revolute joints, the axis of translation for prismatic joints, and the surface normal for planar joints. The axis is specified in the joint frame of reference. Fixed and floating joints do not use the axis field.

| attribute | type     | use      | default value | description                                                                       |
| --------- | -------- | -------- | ------------- | --------------------------------------------------------------------------------- |

> | `xyz`     | `string` | required | NA            | Represents the (x, y, z) components of a vector. The vector should be normalized. |

Reviewer: ```suggestion
| `xyz`     | `string` | required for some joint types | NA            | Represents the (x, y, z) components of a vector. The vector should be normalized. |
```

Author: @scpeters @sloretz let me know if you like this change in, thanks!

### `<calibration>`

The reference positions of the joint, used to calibrate the absolute position of the joint.

| attribute | type     | use      | default value | description                                                                                        |
| --------- | -------- | -------- | ------------- | -------------------------------------------------------------------------------------------------- |
| `rising`  | `double` | optional | ???           | When the joint moves in a positive direction, this reference position will trigger a rising edge.  |
| `falling` | `double` | optional | ???           | When the joint moves in a positive direction, this reference position will trigger a falling edge. |

### `<dynamics>`

> An element specifying physical properties of the joint. These values are used to specify modeling properties of the joint, particularly useful for simulation.

Reviewer: ```suggestion
An element specifying physics-related properties of the joint. These values are used to specify modelling properties of the joint, particularly useful for simulation.
```

Author: @scpeters @sloretz let me know if you like this change in, thanks!

| attribute  | type     | use      | default value | description                                                                                                                                                             |
| ---------- | -------- | -------- | ------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `damping`  | `double` | optional | defaults to 0 | The physical damping value of the joint (in newton-seconds per metre [N∙s/m] for prismatic joints, in newton-metre-seconds per radian [N∙m∙s/rad] for revolute joints). |
| `friction` | `double` | optional | defaults to 0 | The physical static friction value of the joint (in newtons [N] for prismatic joints, in newton-metres [N∙m] for revolute joints).                                      |

### `<limit>`

> (required only for revolute and prismatic joint)

Reviewer: ```suggestion
(applies only to revolute and prismatic joints)
```

| attribute  | type     | use      | default value | description                                                                                                                                                                                                                                |
| ---------- | -------- | -------- | ------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |

> | `effort`   | `double` | required | NA            | An attribute for enforcing the maximum joint effort (\|applied effort\| < \|effort\|). See safety limits.                                                                                                                                  |

Reviewer: ```suggestion
| `effort`   | `double` | required | NA            | An attribute for enforcing the maximum joint effort in Newtons or Newton-meters (\|applied effort\| < \|effort\|). See safety limits.                                                                                                                                  |
```

Author: @scpeters @sloretz let me know if you like this change in, thanks!

| `multiplier` | `double` | optional | defaults to 1 | Specifies the multiplicative factor in the formula above.                                                                  |
| `offset`     | `double` | optional | defaults to 0 | Specifies the offset to add in the formula above. Defaults to 0 (radians for revolute joints, meters for prismatic joints) |

### `<safety_controller>`

| attribute          | type     | use      | default value | description                                                                                                                                                                                                                                                                                    |
| ------------------ | -------- | -------- | ------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `k_velocity`       | `double` | required | NA            | An attribute specifying the relation between effort and velocity limits. See See safety limits for more details.                                                                                                                                                                               |

Reviewer: ```suggestion
| `k_velocity`       | `double` | required | NA            | An attribute specifying the relation between effort and velocity limits. See safety limits for more details.                                                                                                                                                                               |
```

Reviewer 2: Agreed we should take this fix now.
